README.md
# <a name="top"></a>xAPI

* [概述](#概述)
* [安装](#安装)
* [Demo](#demo)
* [使用](#使用)
   * [Objc](#objc)
   * [Swift](#swift)
   * [自适应服务端返回数据结构](#adaptret)
* [公共参数](#公共参数)
   * [commonParams](#commonparams)
   * [commonHeaders](#commonheaders)
   * [区分不同子系统](#tags)

## 概述

业务无关的链式API调用工具，支持 objc / swift

底层网络访问基于 [AFNetworking](https://github.com/AFNetworking/AFNetworking)

异步返回处理基于google的 [promise](https://github.com/google/promises)

## 安装

通过pod引用，在podfile增加下面一行，通过tag指定版本
```
pod 'xAPI',         :git => "https://github.com/jinsikui/xAPI.git", :tag => 'vX.X.X-X'
```
 objc代码中引入：
```
#import <xAPI/xAPI.h>
```
 swift代码中引入：
```
在项目的plist文件中配置置 'Objective-C Bridging Header = xxx.h'，在xxx.h中添加一行：
#import <xAPI/xAPI.h>
```

## <a name="demo"></a> Demo

直接下载本工程运行即可启动demo

## 使用

### <a name="objc"></a> - Objc 

```objc
xAPI.host(@"https://xxx.com")
   .method(HTTP_GET)
   .path([NSString stringWithFormat:@"/v1/%lu/users", (unsigned long)roomId])
   .execute().then(^id(NSDictionary *ret){
      // process api ret in main thread
      NSArray *users = ret[@"users"];
      // do business logic ...
      return nil;
   }).catch(^(NSError *error){
      // 错误处理
      NSInteger code = error.code;
      NSString *errorMsg = error.userInfo[xAPIErrorMessageKey]; //errorMsg could be @""
      // do with error info ...
   });

```
### <a name="swift"></a> - Swift

```swift
xAPI.host("https://xxx.com").method(xHTTPMethod.HTTP_GET)
   .path("/v1/\(roomId)/users")
   .execute().__onQueue(DispatchQueue.main, then:{ ret in
        // process api ret in main thread
        // 这里提取数据的类型转换有点繁琐，推荐用 SwiftyJSON 等三方库来处理json对象
        let dic = ret as! [String : AnyObject]
        let users = dic["users"] as! Array<[String : AnyObject]>
        // do business logic ...
        return nil
   }).__onQueue(DispatchQueue.main, catch:{ error in
        // 错误处理
        let error = error as NSError
        let code = error.code
        let errorMsg = error.userInfo[xAPIErrorMessageKey] //errorMsg could be ""
        // do with error info ...
   })
```

### <a name="adaptret"></a> - 自适应服务端返回数据结构

```objc
// xAPI会对服务端返回的原始数据中的"data","error code","error message"三部分信息进行提取，并且对这几部分兼容不同的key，以适配不同的服务端团队
// xAPI内部代码：
    NSInteger errorCode = 0;
    NSString *errorMsg = nil;
    if (response.statusCode != 200) {
        errorCode = response.statusCode > 0 ? response.statusCode : 110; /*110代表请求超时*/
    }
    if ([response.responseObject isKindOfClass:NSDictionary.class]) {
        NSDictionary *body = response.responseObject;
        if (xapi_not_null(body[@"errcode"]) || xapi_not_null(body[@"code"]) || xapi_not_null(body[@"errorno"]) || xapi_not_null(body[@"errCode"])) {
            //code not null
            NSInteger code = xapi_not_null(body[@"errcode"]) ? [body[@"errcode"] integerValue] : xapi_not_null(body[@"code"]) ? [body[@"code"] integerValue] : xapi_not_null(body[@"errorno"]) ?  [body[@"errorno"] integerValue] : [body[@"errCode"] integerValue];
            if (code != 200 && code != 0) {
                errorCode = code;
            }
        }
        errorMsg = xapi_not_null(body[@"errmsg"]) ? body[@"errmsg"] : xapi_not_null(body[@"msg"]) ? body[@"msg"] : xapi_not_null(body[@"errormsg"]) ? body[@"errormsg"] : xapi_not_null(body[@"errMsg"]) ? body[@"errMsg"] : nil;
    }
    if (errorCode > 0) {
        if (completion) {
            completion(response, errorCode, errorMsg);
        }
        return;
    }

    if([response.responseObject isKindOfClass:NSDictionary.class]){
        NSDictionary *body = response.responseObject;
        id data = xapi_not_null(body[@"data"]) ? body[@"data"] : xapi_not_null(body[@"ret"]) ? body[@"ret"] : nil;
        if (data != nil) {
            response = [[xNetworkResponse alloc] initWithResponse:response adpatedObject:data];
        }
    }
    if (completion) {
        completion(response, errorCode, errorMsg);
    }
   // ......
```
## 公共参数

```
对于所有api请求都附带的公共参数，不必每次调用都写一次，xAPI可以配置几个全局的回调(blocks)，在每次发起api请求时都会主动去调用，来获取公共的参数
公共参数包括 commonParams 和 commonHeaders
```

### <a name="commonparams"></a> - commonParams 

```objc
/******************************************************************************************************************************
注意：
1. xAPI.commonParams(...)是一个全局的配置，只需要在app启动时配置一次

2. xAPI.commonParams(...)配置的是一个返回字典的block，而不是字典本身，每次请求api前，这个block都会被调用，用来实时生成公共参数字典

3. 每次api调用都可以链式的指定一个commonParamsUsage，它有两个取值：
   xAPICommonParamsUsagePath：公共参数应用到api请求的querystring中，不论请求用的是不是GET方法都会添加至querystring
   xAPICommonParamsUsageParams：公共参数和api的params合并，不指定encodingType时对于GET请求添加到querystring中，否则添加到请求body的json中
   
4. commonParamUsage不是全局的，而是跟每个api绑定的，默认值为 xAPICommonParamsUsagePath
*******************************************************************************************************************************/

// app启动时配置
xAPI.commonParams(^NSDictionary * _Nonnull(NSString * _Nullable tag) {
   return [System1API commonParams];
});

// 一次具体的api调用
xAPI.host(@"https://xxx.com")
   .path(@"/test")
   .params(@{"p1": p1, @"p2": p2})
   .commonParamUsage(xAPICommonParamsUsageParams) //指定公共参数和params合并
   .method(HTTP_POST)
   .execute()
```
### <a name="commonheaders"></a> - commonHeaders

```objc
/******************************************************************************************************************************
注意：
1. xAPI.commonHeaders(...)是一个全局的配置，只需要在app启动时配置一次

2. xAPI.commonHeaders(...)配置的是一个返回字典的block，而不是字典本身，每次请求api前，这个block都会被调用，用来实时生成公共headers字典

3. 每次api调用也可以通过xAPI.headers(...)指定和这次api请求绑定的headers，会和commonHeaders返回的字典合并
*******************************************************************************************************************************/

// app启动时配置
xAPI.commonHeaders(^NSDictionary * _Nonnull(NSString * _Nullable tag) {
   return [System1API commonHeaders];
});

// 一次具体的api调用
xAPI.host(@"https://xxx.com")
   .path(@"/auth")
   .headers(@{"SECRET": secret}) // 本次请求绑定的headers
   .method(HTTP_GET)
   .execute()
```

### <a name="tags"></a> - 区分不同子系统

```
如上所述，xAPI.commonParams(...) 和 xAPI.commonHeaders(...) 都是全局的配置，都只能配置一个，如果不同的子系统要求的公共参数不一样怎么办呢？
这可以通过tag来解决，每次具体的api请求都可以链式的指定一个tag，这个tag会作为参数传给 xAPI.commonParams(...) 和 xAPI.commonHeaders(...)配置的block：
```
```objc

// in file AppDelegate.m ...
// ......

// 全局的公共params
xAPI.commonParams(^NSDictionary * _Nonnull(NSString * _Nullable tag) {
   if([tag isEqualToString:System1API.tag]){
      // 子系统1
      return [System1API commonParams];
   }
   else if([tag isEqualToString:System2API.tag]){
      // 子系统2
      return [System2API commonParams];
   }
   else {
      // 其他子系统
      return @{};
   }
});

// 全局的公共headers
xAPI.commonHeaders(^NSDictionary * _Nonnull(NSString * _Nullable tag) {
   if([tag isEqualToString:System1API.tag]){
      // 子系统1
      return [System1API commonHeaders];
   }
   else if([tag isEqualToString:System2API.tag]){
      // 子系统2
      return [System2API commonHeaders];
   }
   else {
      // 其他子系统
      return @{};
   }
});


// ......
// in file System1API.m（子系统1的所有API都放在这里）

/// 子系统公共逻辑
-(xAPIBuilder*)builder{
    return xAPI.tag(System1API.tag).host(System1API.host);
}

/// 某个具体的API接口
-(FBLPromise<NSDictionary*>*)getUsersInRoom:(NSString *)roomId {
    return self.builder.method(HTTP_GET).path([NSString stringWithFormat:@"/v1/%@/users", roomId]).execute();
}


// ......
// in file System2API.m（子系统2的所有API都放在这里）

/// 子系统公共逻辑
-(xAPIBuilder*)builder{
    return xAPI.tag(System2API.tag).host(System2API.host);
}

/// 某个具体的API接口
-(FBLPromise<NSDictionary*>*)getAudiencesInRoom:(NSUInteger)roomId{
    return self.builder.method(HTTP_GET).path([NSString stringWithFormat:@"/v1/%lu/audiences", (unsigned long)roomId]).execute();
}

```




CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [2.1.0-0](https://github.com/jinsikui/xAPI/compare/v2.0.0-0...v2.1.0-0) (2022-04-03)


### Bug Fixes

* podspec ([be950d7](https://github.com/jinsikui/xAPI/commit/be950d797fe9530ab77c55d116304cddda8993ad))

## [2.0.0-0](https://github.com/jinsikui/xAPI/compare/v1.0.0-1...v2.0.0-0) (2022-04-03)


### Features

* code complete ([14171b7](https://github.com/jinsikui/xAPI/commit/14171b7c319794ec916f7243c804e262aecc4b44))
* readme ([8004073](https://github.com/jinsikui/xAPI/commit/8004073c35c06f51c378da032ac79e538c77e499))

## 1.0.0-1 (2022-03-30)


### Features

* setup project ([237d002](https://github.com/jinsikui/xAPI/commit/237d00295e04f6a937ee0fdd6091d0b4be705dcf))


Source/Classes/Services/xAPI.m


#import "xAPI.h"

@interface xAPIBaseBuilder ()
@property (nonatomic, copy, nullable, readwrite) SAPISetParamsBlock paramsBlock;
@property (nonatomic, copy, nullable, readwrite) SAPISetHeadersBlock headersBlock;
@property (nonatomic, copy) NSString *mDefaultHost;
@property (nonatomic, assign) xHTTPMethod mDefaultMethod;
@end

@implementation xAPIBaseBuilder

+ (instancetype)shared {
    static xAPIBaseBuilder *instance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        instance = [[xAPIBaseBuilder alloc] init];
    });
    return instance;
}

- (void (^)(SAPISetParamsBlock _Nonnull))commonParams {
    return ^(SAPISetParamsBlock commonParamsBlock) {
        self.paramsBlock = commonParamsBlock;
    };
}

- (void (^)(SAPISetHeadersBlock _Nonnull))commonHeaders {
    return ^(SAPISetHeadersBlock commonHeadersBlock) {
        self.headersBlock = commonHeadersBlock;
    };
}

- (void (^)(NSString *defaultHost))defaultHost {
    return ^(NSString *defaultHost) {
        self.mDefaultHost = [defaultHost copy];
    };
}

- (void (^)(xHTTPMethod defaultMethod))defaultMethod {
    return ^(xHTTPMethod defaultMethod) {
        self.mDefaultMethod = defaultMethod;
    };
}

@end

@interface xAPIConvertable () {
    NSDictionary *_mParams;
    NSDictionary *_mHeaders;
    NSString *_mHost;
    NSString *_mPath;
    xHTTPMethod _mMethod;
    xParameterEncoding _mEncodingType;
    xResponseDecncoding _mDecodingType;
    xNetworkSub *_mStubData;
}
@end
@implementation xAPIConvertable

- (instancetype)initWithParams:(NSDictionary *)params
                       headers:(NSDictionary *)headers
                          host:(NSString *)host
                          path:(NSString *)path
                        method:(xHTTPMethod)method
                  encodingType:(xParameterEncoding)encodingType
                  decodingType:(xResponseDecncoding)decodingType
                      stubData:(xNetworkSub*)stubData{
    self = [super init];
    if (self) {
        _mParams = params;
        _mHeaders = headers;
        _mHost = host;
        _mPath = path;
        _mMethod = method;
        _mEncodingType = encodingType;
        _mDecodingType = decodingType;
        _mStubData = stubData;
    }
    return self;
}

#pragma mark - xNetworkConvertable
- (NSString*)baseURL {
    return _mHost;
}

- (NSDictionary<NSString *,NSString *> *)httpHeaders {
    return _mHeaders;
}

- (NSString *)path {
    return _mPath;
}

- (xHTTPMethod)httpMethod {
    return _mMethod;
}

- (NSDictionary *)parameters {
    return _mParams;
}

- (xParameterEncoding)encodingType {
    return _mEncodingType;
}

- (xResponseDecncoding)decodingType {
    return _mDecodingType;
}

- (xNetworkSub*)stubData{
    return _mStubData;
}

- (xNetworkResponse *)adaptResponse:(xNetworkResponse *)networkResponse {
    return [[xNetworkResponse alloc] initWithResponse:networkResponse adpatedObject:networkResponse.responseObject];
}

@end

@interface xAPIBuilder ()

@property (nonatomic, copy, readwrite) NSString *mTag;
@property (nonatomic, strong, readwrite) xNetworkSub *mStubData;
@property (nonatomic, strong, readwrite) NSDictionary *mParams;
@property (nonatomic, strong, readwrite) NSDictionary *mHeaders;
@property (nonatomic, assign, readwrite) xAPICommonParamsUsage mParamUsage;
@property (nonatomic, assign, readwrite) xAPICommonHeadersUsage mHeaderUsage;
@property (nonatomic, copy, readwrite) NSString *mHost;
@property (nonatomic, copy, readwrite) NSString *mPath;
@property (nonatomic, assign, readwrite) xHTTPMethod mMethod;
@property (nonatomic, assign, readwrite) xParameterEncoding mEncodingType;
@property (nonatomic, assign, readwrite) xResponseDecncoding mDecodingType;
@property (nonatomic, copy, nullable, readwrite) SAPIBeforeNetworkCallbackBlock beforeCallback;

@property (strong, nonatomic) xNetworkManager * manager;
@end

@implementation xAPIBuilder

- (instancetype)init {
    self = [super init];
    if (self) {
        self.manager = [xNetworkManager manager];
        self.mEncodingType = xParameterEncodingJSON;
        self.mDecodingType = xResponseDecncodingJSON;
        self.mParamUsage = xAPICommonParamsUsagePath;
        self.mHeaderUsage = xAPICommonHeadersUsageHeaders;
        self.mMethod = [xAPIBaseBuilder shared].mDefaultMethod;
        self.mHost = [xAPIBaseBuilder shared].mDefaultHost;
    }
    return self;
}

- (xAPIBuilder * _Nonnull (^)(NSString * _Nonnull))tag {
    return ^id(NSString *tag) {
        self.mTag = tag;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(xNetworkSub * _Nonnull))stubData {
    return ^id(xNetworkSub *stubData) {
        self.mStubData = stubData;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(NSDictionary * _Nonnull))params {
    return ^id(NSDictionary *params) {
        self.mParams = params;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(NSDictionary * _Nonnull))headers {
    return ^id(NSDictionary *headers) {
        self.mHeaders = headers;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(xAPICommonParamsUsage))commonParamsUsage {
    return ^id(xAPICommonParamsUsage usage) {
        self.mParamUsage = usage;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(xAPICommonHeadersUsage))commonHeadersUsage {
    return ^id(xAPICommonHeadersUsage usage) {
        self.mHeaderUsage = usage;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(NSString * _Nonnull))host {
    return ^id(NSString *host) {
        self.mHost = [host copy];
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(NSString * _Nonnull))path {
    return ^id(NSString *path) {
        self.mPath = [path copy];
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(xHTTPMethod))method {
    return ^id(xHTTPMethod method) {
        self.mMethod = method;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(xParameterEncoding))encodingType {
    return ^id(xParameterEncoding encodingType) {
        self.mEncodingType = encodingType;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(xResponseDecncoding))decodingType {
    return ^id(xResponseDecncoding decodingType) {
        self.mDecodingType = decodingType;
        return self;
    };
}

- (xAPIBuilder * _Nonnull (^)(SAPIBeforeNetworkCallbackBlock _Nonnull))beforeNetworkCallback {
    return ^id(SAPIBeforeNetworkCallbackBlock beforeCallback) {
        self.beforeCallback = beforeCallback;
        return self;
    };
}

- (FBLPromise * (^)(void))execute {
    return ^FBLPromise * {
        FBLPromise * (^networkPromise)(void) = ^{
            return FBLPromise.asyncOn(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^(FBLPromiseFulfillBlock fulfill, FBLPromiseRejectBlock reject) {
                void (^myCompletion)(xNetworkResponse * _Nonnull, NSInteger, NSString*) = ^(xNetworkResponse * _Nonnull response, NSInteger errorCode, NSString *errorMsg) {
                    if (errorCode != 0) {
                        NSError *error = [NSError errorWithDomain:xAPIErrorDomain code:errorCode userInfo:@{xAPIErrorMessageKey: errorMsg ?: @""}];
                        reject(error);
                    } else {
                        fulfill(response.responseObject);
                    }
                };
                [self.manager request:[self createConvertable] completion:[self getRequestCompletionFromCompletion:myCompletion]];
            });
        };
        /// 是否有在网络请求之前的回调
        if (self.beforeCallback) {
            FBLPromise *beforePromise = self.beforeCallback();
            if (beforePromise) {
                return beforePromise.then(^id _Nullable(id  _Nullable value) {
                    return networkPromise();
                });
            }
        }
        return networkPromise();
    };
}

- (xAPIConvertable *)createConvertable {
    NSMutableDictionary *commonParams = [NSMutableDictionary dictionary];
    if ([xAPIBaseBuilder shared].paramsBlock) {
        commonParams = [[xAPIBaseBuilder shared].paramsBlock(self.mTag) mutableCopy];
    }
    NSMutableDictionary *commonHeaders = [NSMutableDictionary dictionary];
    if ([xAPIBaseBuilder shared].headersBlock) {
        commonHeaders = [[xAPIBaseBuilder shared].headersBlock(self.mTag) mutableCopy];
    }
    /// 参数
    if (self.mParamUsage & xAPICommonParamsUsageParams) {
        [commonParams addEntriesFromDictionary:self.mParams];
        self.mParams = commonParams;
    }
    /// 头
    if (self.mHeaderUsage & xAPICommonHeadersUsageHeaders) {
        [commonHeaders addEntriesFromDictionary:self.mHeaders];
        self.mHeaders = commonHeaders;
    }
    /// 地址
    if (self.mParamUsage & xAPICommonParamsUsagePath) {
        self.mPath = [xAPIHelper mergeToInput:self.mPath queryParams:commonParams];
    }
    
    /// 编码
    xParameterEncoding encodingType = self.mMethod == HTTP_GET ? xParameterEncodingHTTP : self.mEncodingType;
    
    return [[xAPIConvertable alloc] initWithParams:self.mParams
                                           headers:self.mHeaders
                                              host:self.mHost
                                              path:self.mPath
                                            method:self.mMethod
                                      encodingType:encodingType
                                      decodingType:self.mDecodingType
                                          stubData:self.mStubData];
}

- (void(^)(xNetworkResponse * _Nonnull))getRequestCompletionFromCompletion:(void (^)(xNetworkResponse * _Nonnull, NSInteger errorCode, NSString *errorMsg))completion {
    void (^requestCompletion)(xNetworkResponse * _Nonnull) = ^(xNetworkResponse *response) {
        NSInteger errorCode = 0;
        NSString *errorMsg = nil;
        if (response.statusCode != 200) {
            errorCode = response.statusCode > 0 ? response.statusCode : 110; /*110代表请求超时*/
        }
        if ([response.responseObject isKindOfClass:NSDictionary.class]) {
            NSDictionary *body = response.responseObject;
            if (xapi_not_null(body[@"errcode"]) || xapi_not_null(body[@"code"]) || xapi_not_null(body[@"errorno"]) || xapi_not_null(body[@"errCode"])) {
                //code not null
                NSInteger code = xapi_not_null(body[@"errcode"]) ? [body[@"errcode"] integerValue] : xapi_not_null(body[@"code"]) ? [body[@"code"] integerValue] : xapi_not_null(body[@"errorno"]) ?  [body[@"errorno"] integerValue] : [body[@"errCode"] integerValue];
                if (code != 200 && code != 0) {
                    errorCode = code;
                }
            }
            errorMsg = xapi_not_null(body[@"errmsg"]) ? body[@"errmsg"] : xapi_not_null(body[@"msg"]) ? body[@"msg"] : xapi_not_null(body[@"errormsg"]) ? body[@"errormsg"] : xapi_not_null(body[@"errMsg"]) ? body[@"errMsg"] : nil;
        }
        if (errorCode > 0) {
            if (completion) {
                completion(response, errorCode, errorMsg);
            }
            return;
        }
        
        if([response.responseObject isKindOfClass:NSDictionary.class]){
            NSDictionary *body = response.responseObject;
            id data = xapi_not_null(body[@"data"]) ? body[@"data"] : xapi_not_null(body[@"ret"]) ? body[@"ret"] : nil;
            if (data != nil) {
                response = [[xNetworkResponse alloc] initWithResponse:response adpatedObject:data];
            }
        }
        if (completion) {
            completion(response, errorCode, errorMsg);
        }
    };
    return requestCompletion;
}

@end

@implementation xAPI

#pragma mark - Common API Properties

+ (void)buildCommon:(void (^)(xAPIBaseBuilder * _Nonnull baseBuilder))block {
    block([xAPIBaseBuilder shared]);
}

+ (void (^)(SAPISetParamsBlock _Nonnull))commonParams {
    return xAPIBaseBuilder.shared.commonParams;
}

+ (void (^)(SAPISetHeadersBlock _Nonnull))commonHeaders {
    return xAPIBaseBuilder.shared.commonHeaders;
}

+ (void (^)(NSString * _Nonnull))defaultHost {
    return xAPIBaseBuilder.shared.defaultHost;
}

+ (void (^)(xHTTPMethod))defaultMethod {
    return xAPIBaseBuilder.shared.defaultMethod;
}

#pragma mark - Instance API Properties

+ (xAPIBuilder *)build:(void (^)(xAPIBuilder * _Nonnull))block {
    xAPIBuilder *builder = [[xAPIBuilder alloc] init];
    if (block) {
        block(builder);
    }
    return builder;
}

+ (xAPIBuilder * _Nonnull (^)(NSString * _Nonnull))tag {
    return [[[xAPIBuilder alloc] init] tag];
}

+ (xAPIBuilder * _Nonnull (^)(xNetworkSub * _Nonnull))stubData {
    return [[[xAPIBuilder alloc] init] stubData];
}

+ (xAPIBuilder * _Nonnull (^)(NSDictionary * _Nonnull))params {
    return [[[xAPIBuilder alloc] init] params];
}

+ (xAPIBuilder * _Nonnull (^)(NSDictionary * _Nonnull))headers {
    return [[[xAPIBuilder alloc] init] headers];
}

+ (xAPIBuilder * _Nonnull (^)(xAPICommonParamsUsage))commonParamsUsage {
    return [[[xAPIBuilder alloc] init] commonParamsUsage];
}

+ (xAPIBuilder * _Nonnull (^)(xAPICommonHeadersUsage))commonHeadersUsage {
    return [[[xAPIBuilder alloc] init] commonHeadersUsage];
}

+ (xAPIBuilder * _Nonnull (^)(NSString * _Nonnull))host {
    return [[[xAPIBuilder alloc] init] host];
}

+ (xAPIBuilder * _Nonnull (^)(NSString * _Nonnull))path {
    return [[[xAPIBuilder alloc] init] path];
}

+ (xAPIBuilder * _Nonnull (^)(xHTTPMethod))method {
    return [[[xAPIBuilder alloc] init] method];
}

+ (xAPIBuilder * _Nonnull (^)(xParameterEncoding))encodingType {
    return [[[xAPIBuilder alloc] init] encodingType];
}

+ (xAPIBuilder * _Nonnull (^)(SAPIBeforeNetworkCallbackBlock _Nonnull))beforeNetworkCallback {
    return [[[xAPIBuilder alloc] init] beforeNetworkCallback];
}

@end
